### ISSUE TITLE
【2025/04】CAT: Circular-Convolutional Attention for Sub-Quadratic Transformers

### ISSUE THREAD

**Issue by personabb** [personabb/survey_paper#48]
## 論文タイトル（原文まま）
CAT: Circular-Convolutional Attention for Sub-Quadratic Transformers

## 一言でいうと
TransformerのAttention計算量を、FFTベースの循環畳み込みを用いることで、softmax構造を維持したままO(N^2)からO(N log N)に削減し、性能を維持・向上させる新しいAttention機構「CAT」を提案。

## 論文リンク
https://arxiv.org/abs/2504.06704v1

## 著者/所属機関
Yoshihiro Yamada / Preferred Networks Tokyo, Japan

## 投稿日付(yyyy/MM/dd)
2025/04/09

### In this paper,
TransformerのAttention機構が持つO(N^2)の計算量問題を解決するため、softmaxベースの重み付けという重要な特性を維持しつつ計算量を削減する「Engineering-Isomorphism」という設計フレームワークを提唱する。その具体的な実現例として、高速フーリエ変換(FFT)ベースの循環畳み込みを活用した「CAT (Circular-convolutional ATtention)」を提案する。CATはO(N log N)の計算量を達成し、全結合層を合理化することで学習可能なパラメータ数を削減し、追加の重い演算を導入しないという特徴を持つ。

### As a result,
ImageNet-1kやWikiText-103などの大規模ベンチマークにおいて、CATは標準的なAttentionと比較して一貫して精度を向上させ、単純なPyTorch実装においても約10%の速度向上を達成した。提案されたEngineering-Isomorphismフレームワークとその具体例であるCATは、実用的な効率性と実装の容易さを提供するだけでなく、次世代の高性能Transformerアーキテクチャ開発に向けた洞察を与える。さらに、アブレーション研究を通じてCATの有効性の根底にある重要な条件を明らかにし、スケーラブルなAttention機構に関するより広範な原則に光を当てた。

## 先行研究と比べてどこがすごい？
既存のAttention計算量削減手法には以下のような課題があった。
*   **線形Attention (Performer, Linear Transformerなど)**: O(N)計算量を達成するが、softmaxをカーネル関数で近似するため、元のsoftmax構造が失われ、学習の不安定性や精度低下を招くことがある。
*   **状態空間モデル (S4, Mambaなど)**: Attentionとは根本的に異なる機構を採用し、Transformerの設計から大きく逸脱するため、パラメータ数の増加や他の部分での複雑化を招き、単純なドロップイン置換とはなりにくい。
*   **プーリング/混合ベース (PoolFormer, MLP-Mixerなど)**: 計算量は削減されるが、Attentionが持つトークン間の柔軟でデータ依存な相互作用モデリング能力を犠牲にする可能性がある。
*   **スパース/低ランクAttention (BigBird, Longformerなど)**: O(N^2)未満を達成するが、大域的なsoftmax重み付けを部分的にしか適用しないため、Engineering-Isomorphismの定義（特にSoftmax Preservation）を満たさない。また、しばしばシーケンス長Nに依存するハイパーパラメータ（ブロックサイズなど）の調整が必要となり、Minimal Hyperparameter Overheadの原則に反する。

これに対し、CATは「Engineering-Isomorphism」の原則（①Softmax Preservation, ②Sub-Quadratic Complexity, ③Comparable or reduced parameter count, ④Minimal Hyperparameter Overhead）に従うことで、以下の点で優れている。
*   **Softmax構造の維持**: 大域的なsoftmaxによる重み付け構造を保持するため、Attentionの表現力や解釈可能性を維持し、学習の安定性が期待できる。
*   **計算量の削減**: FFTを用いることで、O(N log N)の計算量を達成する。
*   **パラメータ効率**: QueryとKeyの射影を統合することで、標準的なAttentionよりもパラメータ数を削減できる。
*   **少ない追加要素**: Nに依存する複雑なハイパーパラメータや、重い演算を追加しない。

これにより、性能と安定性を損なわずにスケーラビリティを向上させ、既存Transformerへの統合も容易である。

## 技術や手法のキモはどこ？
1.  **Engineering-Isomorphismフレームワーク**: softmaxベースのAttentionの本質を維持しつつ、計算量をO(N^2)未満に削減するための4つの条件を定義した点。
    1.  **Softmax Preservation**: `softmax(F_attn(x)) * F_value(x)` の形式を保持する。
    2.  **Sub-Quadratic Complexity**: O(N^2)未満で計算可能。
    3.  **Comparable or reduced parameter count**: パラメータ数が標準Attention以下。
    4.  **Minimal Hyperparameter Overhead**: Nに依存する追加ハイパーパラメータが少ない。
2.  **循環畳み込み (Circular Convolution)**: 標準Attentionの `QK^T` 行列計算を、循環行列を用いた操作で置き換える。具体的には、入力 `x` から射影したベクトル `z = xWA` を計算し、`softmax(z)` の結果を用いて循環行列 `Roll(softmax(z))` を（暗黙的に）構築する。この循環構造により、各トークンが他の全てのトークンと相互作用する大域的な関係性を保持する。
3.  **FFTによる高速化**: 循環行列とValueベクトル `V = xWV` の積 `Roll(softmax(z)) * V` は、畳み込み定理により、FFTを用いて周波数領域での要素ごとの積として計算できる。これにより、計算量が O(N^2) から O(N log N) に削減される (`F_cat = IFFT( FFT(softmax(xWA)) * FFT(xWV) )`)。
4.  **パラメータ効率化**: Query (`Q`) と Key (`K`) の射影を行わず、代わりに単一の射影 `WA` を学習する (`qv` アプローチ)。これにより、標準Attentionの `WQ, WK, WV` (計 `3d^2`) よりも少ないパラメータ数 (`WA` は `d*1`、`WV` は `d*d` なので、マルチヘッドを考慮しても削減される。論文Tableでは `(d+h)d`) で同等の機能を実現しようとする。

## どうやって有効だと検証した？
画像分類タスク (ImageNet-1k) と言語モデリングタスク (WikiText-103) の大規模ベンチマークを用いてCATの有効性を検証した。
*   **モデル**:
    *   画像分類: ViT CLIP-B/L
    *   言語モデリング: Transformer-XL, GPT-2 small
*   **比較対象**:
    *   標準的なAttention (ベースライン)
    *   CAT (全てのAttention層をCATに置き換え)
    *   CAT-Alter (半分のAttention層をCATに置き換え、残りは標準Attention)
*   **評価指標**:
    *   ImageNet-1k: 分類精度 (Accuracy)
    *   WikiText-103: 単語パープレキシティ (word PPL)
*   **実験設定**: ベースラインモデルのハイパーパラメータ（学習率、バッチサイズ、エポック数、オプティマイザなど）を継承して学習。画像分類ではtoken poolingとaverage poolingの2種類を評価。言語モデリングではmasked LMとcausal LMの両方を評価（causal LMでは未来のトークンを見ないように循環シフトを調整）。
*   **主な結果**:
    *   **ImageNet-1k**: Average Poolingを用いた場合にCATは標準Attentionを上回る精度を示した。CAT-Alterはtoken/average poolingの両方で頑健であり、多くの場合で標準Attentionよりも高い精度を達成した (例: CLIP-L avg poolingでAttention 0.646に対しCAT 0.694, CAT-Alter 0.681)。
    *   **WikiText-103**: Masked LMにおいてCATおよびCAT-Alterは標準Attentionより大幅にPPLを改善した (例: Transformer-XLでAttention 13.94に対しCAT 10.28, CAT-Alter 8.51)。Causal LMにおいては、CAT-Alterが標準Attentionと同等以上の性能を示した (例: GPT-2 smallでAttention 27.84に対しCAT-Alter 27.68)。
    *   **速度**: 単純なPyTorch実装 (torch.gatherを用いたもの。FFT実装は同等か僅かに速い程度) で、N=256、V100 GPU上で標準Attentionと比較して約10%の高速化を確認した。
    *   **線形Attention**: 比較のために試したが、学習が不安定 (NaN loss) になり、安定した結果が得られなかったため、主要な比較からは除外した。

これらの結果から、CATは計算量を削減しつつ、特にAverage PoolingやMasked LMといった設定で標準Attentionと同等以上の性能を発揮できること、CAT-Alterはより広範な設定で有効な選択肢であることが示された。

## 議論はある？
*   **CATの有効性が高い場面**: 実験結果から、CATはAverage PoolingやMasked Language Modelingなど、トークン間の相互作用が比較的単純化される、あるいは大域的な情報集約が重視される場面で特に高い性能を発揮する傾向がある。これは、近年のVision TransformerでAverage Poolingが好まれる傾向や、MLMの重要性と合致しており、CATの有用性を示唆している。
*   **CAT-Alterの有効性**: 全ての層を置き換えるのではなく、一部の層のみをCATにするCAT-Alterが、多くの場合で標準Attentionと同等以上の性能を達成し、より頑健な選択肢となることが示された。これは、効率性と表現力のバランスを取るため現実的なアプローチと言える。
*   **パラメータ設計 (Ablation Study)**: Query/Key/Valueの射影方法について比較検討した結果、QueryとKeyの射影を統合する `qv` (CATのデフォルト) アプローチが、パラメータ数と性能のバランスにおいて最も優れていることが示された。完全な `qkv` (Averaged-Key) はわずかに性能が高いがパラメータ数が増え、`q` のみや `v` のみでは性能が低下した。
*   **Mambaとの関連性**: `qv` アプローチがQuery様の情報をValueの計算に組み込む形になっている点は、Mambaのような状態空間モデルに見られる入力依存の選択的状態更新メカニズムと概念的に類似している可能性が示唆された。これが `qv` アプローチの性能の高さに寄与しているかもしれない。
*   **実装と効率**: 理論上O(N log N)だが、現状のPyTorch実装ではgatherベースの方がFFTベースより同等か僅かに速い場合がある。FFTの効率はシーケンス長Nやハードウェアに依存するため、超長系列や専用カーネルでの検証が重要。FlashAttentionのようなメモリ効率化技術との組み合わせも有望である。
*   **今後の課題**:
    *   超長系列 (数万トークンレベル) での学習安定性、精度、実測速度の検証。
    *   他の効率化Attention手法（スパース、低ランク、カーネル法）との組み合わせの探求。
    *   GPUや専用アクセラレータ向けのハードウェア最適化FFTカーネルの開発・適用。
    *   自然言語、画像以外のドメイン（音声、時系列予測など）への応用。

## 結果
*   **ImageNet-1k (ViT CLIP-L, avg pooling, Accuracy ↑)**:
    *   Standard Attention: 0.646
    *   CAT (ours): 0.694 (+0.048)
    *   CAT-Alter (ours): 0.681 (+0.035)
*   **WikiText-103 (Transformer-XL, masked LM, word PPL ↓)**:
    *   Standard Attention: 13.94
    *   CAT (ours): 10.28 (-3.66)
    *   CAT-Alter (ours): 8.51 (-5.43)
*   **WikiText-103 (GPT-2 small, masked LM, word PPL ↓)**:
    *   Standard Attention: 9.82
    *   CAT (ours): 8.32 (-1.50)
    *   CAT-Alter (ours): 7.54 (-2.28)
*   **WikiText-103 (GPT-2 small, causal LM, word PPL ↓)**:
    *   Standard Attention: 27.84
    *   CAT (ours): 32.36 (+4.52)
    *   CAT-Alter (ours): 27.68 (-0.16)
*   **計算量**: O(N log N) (理論値)
*   **速度**: 約10%向上 (Naive PyTorch, N=256, V100 GPU)
*   **パラメータ数**: 標準Attention (`3d^2`) より削減 (CATは `(d+h)d`, CAT-Alterは中間)。
*   **メモリ**: O(N) (FFT実装の場合、N×N行列を生成しないため)。標準AttentionはO(N^2) (Naive実装)。

## 次に読むべき論文は？
*   **Mamba [8, 9] (Gu and Dao, 2023, 2024)**: CATとの比較や、状態空間モデルとAttentionの関係性を深く理解するために。特に[9]はTransformerとSSMの双対性について論じている。
*   **FlashAttention [6] (Dao et al., 2022)**: Attention計算のメモリ効率と速度を大幅に改善する実践的な技術。CATのようなアルゴリズムレベルの効率化と、FlashAttentionのような実装レベルの効率化を組み合わせる可能性を探る上で重要。
*   **Rethinking attention with performers [3] (Choromanski et al., 2021)** や **Transformers are rnns [11] (Katharopoulos et al., 2020)**: 線形Attentionの代表的な研究であり、CATが克服しようとしたsoftmax近似の課題やトレードオフを理解するために。
*   **The hedgehog & the porcupine [23] (Zhang et al., 2024)**: 線形Attentionの表現力と安定性の問題、特にsoftmax模倣について論じており、CATがsoftmax構造を維持することの重要性を理論的に補強する可能性がある。
*   **Longformer [1] (Beltagy et al., 2020)** や **Big Bird [22] (Zaheer et al., 2020)**: スパースAttentionによる効率化アプローチの代表例。CATとは異なる設計思想（局所的Attention + 大域的トークンなど）との比較検討のために。

## コメント
「Engineering-Isomorphism」という設計原則を明確に打ち出し、それに合致する具体的な手法としてCATを提案している点が興味深い。FFTを用いた畳み込みによる計算量削減は過去にもアイデアがあったかもしれないが、それをTransformerの核であるsoftmax構造を維持しつつ、さらにパラメータ削減も同時に達成しようとしている点が本研究の独自性であり、評価できるポイントだと考える。特に、全てのAttention層を置き換える「CAT」だけでなく、部分的に標準Attentionを残す「CAT-Alter」が高い性能と頑健性を示すという結果は、既存モデルへの適用や性能と効率のバランスを取る上で現実的で示唆に富む。Ablation Studyで `qv` 方式の有効性を示し、Mambaとの関連性にまで言及している点も面白い。実装の詳細、特にgatherベースとFFTベースの実装の使い分けや、causal maskのための循環シフト調整方法はもう少し具体的に知りたいところである。今後の超長系列データでの実証的な性能評価に期待したい。

## 手法の詳細（数式や理論展開など）
*   **標準Self-Attention**:
    入力 `x ∈ R^(N×D)` に対し、`Q = xW_Q`, `K = xW_K`, `V = xW_V` (`W_Q, W_K, W_V ∈ R^(D×D)`) を計算し、
    `Attention(Q, K, V) = softmax( (QK^T) / sqrt(d_k) ) V`
    ここで `d_k` はKey/Queryの次元（通常 `D/num_heads`）、softmaxは行ごとに適用される。計算量は `QK^T` の O(N^2 D) と `softmax(...)V` の O(N^2 D) がボトルネックとなり、全体で O(N^2 D)。メモリも O(N^2)。

*   **Engineering-Isomorphismの4条件**:
    1.  **Softmax Preservation**: 出力が `F_out(x) = softmax(F_attn(x)) F_value(x)` の形式であること。`F_attn(x)` はAttentionスコアに相当し、`F_value(x)` はValueに相当する。
    2.  **Sub-Quadratic Complexity**: `F_out(x)` の計算量が O(N^2) より厳密に小さいこと。
    3.  **Comparable or reduced parameter count**: 学習パラメータ数が標準Attentionと同等以下であること。
    4.  **Minimal Hyperparameter Overhead**: シーケンス長Nに依存する追加のハイパーパラメータ（スパースパターン、ブロックサイズなど）を導入しないこと。

*   **CAT (Circular-convolutional ATtention)**:
    1.  **Attentionスコアベクトルの計算**: Learnable projection `W_A ∈ R^(D×1)` を用い、`z = x * W_A ∈ R^(N×1)` を計算する。
    2.  **Softmax適用**: `z* = softmax(z) ∈ R^(N×1)` を計算（ベクトル内の全要素でsoftmax）。
    3.  **Valueベクトルの計算**: Learnable projection `W_V ∈ R^(D×D)` を用い、`V = x * W_V ∈ R^(N×D)` を計算する。
    4.  **循環畳み込みによる出力計算**: `z*` から構築される循環行列 `Roll(z*) ∈ R^(N×N)` と `V` の積を計算する。
        `F_cat = Roll(z*) * V`
        ここで `Roll(z*)` は `z*` の要素を循環シフトさせた行列:
        ```
        Roll(z*) = [[z*_1, z*_2, ..., z*_N],
                    [z*_N, z*_1, ..., z*_(N-1)],
                    ...,
                    [z*_2, z*_3, ..., z*_1]]
        ```
        この `Roll(z*)` は陽に計算・保持しない。

    5.  **FFTによる高速化**: 上記の循環行列とベクトルの積は、畳み込み定理を用いてFFTで効率的に計算できる。`z*` と `V` の各次元（チャンネル）に対してFFTを適用し、周波数領域で要素ごとの積を取り、逆FFT (IFFT) を適用する。
        `F_cat = IFFT( FFT(z*) * FFT(V_d) )` for each dimension `d` of `V`.
        実際には、複素数FFTや実数FFT (RFFT) を用いる。これにより計算量は O(N log N) になる。

*   **Averaged-Key (代替案)**:
    標準Attentionに近い形式を保つ代替案。`Q = xW_Q`, `K = xW_K`, `V = xW_V` を計算する。
    Attentionスコアとして、各Query `Q_i` と「平均化されたKey」 `K_avg = (1/N) * sum(K_j)` の内積のようなものを計算する。論文では `z_circ = Q * K_avg` と表現されている (`z_circ ∈ R^N`)。
    その後、`softmax(z_circ)` を計算し、これを `z*` の代わりに用いて `Roll(softmax(z_circ)) * V` を計算する。
    この方法はパラメータ数が標準Attentionに近い (`3d^2`) が、Cross-Attentionへの適用がCAT (`qv`アプローチ) より容易である可能性がある。

*   **Multi-Head**: 標準Attentionと同様に、`W_A` や `W_V` をヘッドごとに独立に用意し、各ヘッドでCAT計算を行った後、結果を結合して最終的な出力を得る。パラメータ数の表記 `(d+h)d` はこれを考慮していると思われる (`h` はヘッド数)。
